File: tests/bindings/java/foo-tests/src/test/java/io/stepfunc/foo_test/PrimitivePointerTest.java
package io.stepfunc.foo_test;

import io.stepfunc.foo.PrimitivePointers;
import io.stepfunc.foo.RangeIteratorTestHelper;
import io.stepfunc.foo.RangeReceiver;
import org.assertj.core.api.Assert;
import org.joou.UInteger;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.joou.Unsigned.ubyte;
import static org.joou.Unsigned.uint;

class PrimitivePointerTest {

    @Test
    void canReadBoolean() {
        PrimitivePointers values = new PrimitivePointers();
        assertThat(values.getBool(true)).isTrue();
        assertThat(values.getBool(false)).isFalse();
    }

    @Test
    void canReadUnsignedByte() {
        PrimitivePointers values = new PrimitivePointers();
        assertThat(values.getU8(ubyte(0))).isEqualTo(ubyte(0));
        assertThat(values.getU8(ubyte(1))).isEqualTo(ubyte(1));
        assertThat(values.getU8(ubyte(254))).isEqualTo(ubyte(254));
        assertThat(values.getU8(ubyte(255))).isEqualTo(ubyte(255));
    }

    @Test
    void canReadFloat() {
        PrimitivePointers values = new PrimitivePointers();
        assertThat(values.getFloat(3.14f)).isEqualTo(3.14f);
        assertThat(values.getFloat(1e6f)).isEqualTo(1e6f);
        assertThat(values.getFloat(0f)).isEqualTo(0f);
    }

    @Test
    void canReadDouble() {
        PrimitivePointers values = new PrimitivePointers();
        assertThat(values.getDouble(3.14)).isEqualTo(3.14);
        assertThat(values.getDouble(1e6)).isEqualTo(1e6);
        assertThat(values.getDouble(0)).isEqualTo(0);
    }

}

File: tests/bindings/dotnet/foo.Tests/PrimitivePointerTest.cs
using System;
using Xunit;
using foo;
using System.Linq;
using System.Collections.Generic;

namespace foo.Tests
{
    public class PrimitivePointrTest
    {
        void AssertFloatEqual(float expected, float value)
        {
            Assert.True(Math.Abs(expected - value) < 1e-6);
        }

        void AssertDoubleEqual(double expected, double value)
        {
            Assert.True(Math.Abs(expected - value) < 1e-6);
        }

        [Fact]
        public void CanReadBoolean()
        {
            var values = new PrimitivePointers();

            Assert.False(values.GetBool(false));
            Assert.True(values.GetBool(true));
        }

        [Fact]
        public void CanReadUnsignedByte()
        {
            var values = new PrimitivePointers();

            Assert.Equal(0, values.GetU8(0));
            Assert.Equal(1, values.GetU8(1));
            Assert.Equal(254, values.GetU8(254));
            Assert.Equal(255, values.GetU8(255));
        }

        [Fact]
        public void CanReadFloat()
        {
            var values = new PrimitivePointers();
            AssertFloatEqual(3.14f, values.GetFloat(3.14f));
            AssertFloatEqual(1e6f, values.GetFloat(1e6f));
            AssertFloatEqual(0f, values.GetFloat(0f));
        }

        [Fact]
        public void CanReadDouble()
        {
            var values = new PrimitivePointers();
            AssertDoubleEqual(3.14, values.GetDouble(3.14));
            AssertDoubleEqual(1e6, values.GetDouble(1e6));
            AssertDoubleEqual(0, values.GetDouble(0));
        }
    }
}
